staging/src/k8s.io/apimachinery/pkg/labels/selector.go

```diff
@@ -963,3 +965,79 @@ func SelectorFromValidatedSet(ls Set) Selector {
 func ParseToRequirements(selector string, opts ...field.PathOption) ([]Requirement, error) {
 	return parse(selector, field.ToPath(opts...))
 }
+
+// SetSelector wraps a Set, allowing it to implement the Selector interface. Unlike
+// Set.AsSelectorPreValidated (which copies the input Set), this type simply wraps the underlying
+// Set. As a result, it is substantially more efficient, but requires the caller to not mutate the
+// Set.
+type SetSelector Set
+
+func (s SetSelector) Matches(labels Labels) bool {
+	for k, v := range s {
+		if !labels.Has(k) || v != labels.Get(k) {
+			return false
+		}
+	}
+	return true
+}
+
+func (s SetSelector) Empty() bool {
+	return len(s) == 0
+}
+
+func (s SetSelector) String() string {
+	keys := make([]string, 0, len(s))
+	for k := range s {
+		keys = append(keys, k)
+	}
+	// Ensure deterministic output
+	sort.Strings(keys)
+	b := strings.Builder{}
+	for i, key := range keys {
+		last := i == len(keys)-1
```

Comment:
can avoid the repeated length check by shifting the comma append to the beginning of the loop when i != 0